nibabel/funcs.py

```diff
@@ -224,7 +224,18 @@ def as_closest_canonical(img, enforce_diag=False):
     # we need to transform the data
     arr = img.get_data()
     t_arr = apply_orientation(arr, ornt)
-    return img.__class__(t_arr, out_aff, img.header)
+    # Also apply the transform to the dim_info fields
+    new_hdr = img.header.copy()
```

Comment:
This is a bit tricky, because the output image does not have to be Nifti type, and therefore the header need not have 'get_dim_info'.

I wonder whether we need a new image method, `transpose`, that deals with this case.  The default implementation could go in `SpatialImage` and be overridden by Nifti.